### PR TITLE
Added new loss functions.

### PR DESCRIPTION
**Pull request by thebhatman**
I have added the KL Divergence Loss function, Poisson loss function, Logcosh loss, and Hinge loss function.

### CODE REVIEW

File: src/layers/stateless.jl [hunk 1]
```diff
@@ -54,3 +54,31 @@ function normalise(x::AbstractArray, dims)
...
+    Poisson Loss function
+Poisson loss function is a measure of how the predicted distribution diverges from the expected distribution.
+
+"""
+Poisson(ŷ, y) = sum(ŷ .- y .* log.(ŷ)) *1 // size(y,2)
+
+"""
+    Logcosh Loss function
+"""
+
```
Comment (MikeInnes) [inline src/layers/stateless.jl]: There are some unnecessary newlines going around here (in the doc strings as well)

**Comment by MikeInnes** [comment]
Seems good; as a point of style it'd be better if the function names were lowercase (we use uppercase for layers / types).

Can you also put a reference for these in the docs?

**Comment by thebhatman** [comment]
Thank you @MikeInnes. I have made the necessary changes.

File: src/layers/stateless.jl [hunk 2]
```diff
@@ -54,3 +54,30 @@ function normalise(x::AbstractArray, dims)
...
+https://isaacchanghau.github.io/post/loss_functions/
+"""
+poisson(ŷ, y) = sum(ŷ .- y .* log.(ŷ)) *1 // size(y,2)
+
+"""
+    Logcosh Loss function
+"""
+logcosh(ŷ, y) = sum(log.(cosh.(ŷ .- y)))
+
+hinge(ŷ, y) = sum(max.(0.0, 1 .-  ŷ .* y)) *1 // size(y,2)
```
Comment (CarloLucibello) [inline src/layers/stateless.jl]: 0.0 -> 0 to avoid promotion to Float64

File: src/layers/stateless.jl [hunk 3]
```diff
@@ -54,3 +54,30 @@ function normalise(x::AbstractArray, dims)
...
+    Poisson Loss function
+Poisson loss function is a measure of how the predicted distribution diverges from the expected distribution.
+https://isaacchanghau.github.io/post/loss_functions/
+"""
+poisson(ŷ, y) = sum(ŷ .- y .* log.(ŷ)) *1 // size(y,2)
+
+"""
+    Logcosh Loss function
+"""
+logcosh(ŷ, y) = sum(log.(cosh.(ŷ .- y)))
```
Comment (CarloLucibello) [inline src/layers/stateless.jl]: this at the same time not very common and easy for users to define, are we sure we want it in Flux?

Comment (thebhatman) [reply src/layers/stateless.jl]: Yes, I agree it is easy to define. But most ML libraries have it predefined and users might expect it to be predefined.

Comment (MikeInnes) [reply src/layers/stateless.jl]: Usually the reason for having it predefined is numerical stability; should we be doing something for that here?

Also, is there a reason for this to not be doing the averaging that other functions do?

Comment (thebhatman) [reply src/layers/stateless.jl]: Thank you @MikeInnes for pointing it out. It was a mistake from my side. logcosh loss has to be averaged. I have corrected it.

Comment (pshashk) [reply src/layers/stateless.jl]: Maybe we could move logcosh activation to NNLib as something like
```
logcosh(x) = x + softplus(-2 * x) - log(2)
∇logcosh(x) = tanh(x)
```
and the loss function will trivially become `logcosh.(ŷ - y)`

Comment (thebhatman) [reply src/layers/stateless.jl]: @MikeInnes would it be better if I remove logcosh function from here if it isn't actually required?

Comment (MikeInnes) [reply src/layers/stateless.jl]: Yeah, it might be better to remove it, as the name doesn't make it obvious that it's meant to be used as a loss.

Comment (thebhatman) [reply src/layers/stateless.jl]: I have removed the logcosh function.

**Comment by thebhatman** [comment]
Where can I add reference to the new loss functions in the docs?

**Comment by MikeInnes** [comment]
Create a section for docstrings [here](https://fluxml.ai/Flux.jl/stable/training/training/#Loss-Functions-1). We should add an API reference at some point, but that'll do for now.

**Comment by pshashk** [comment]
Submitted `logcosh` to NNlib https://github.com/FluxML/NNlib.jl/pull/108/
Should I add `∇logcosh(x) = tanh(x)` gradient to Tracker?

**Comment by thebhatman** [comment]
@MikeInnes, I have added the loss functions to the docs.

**Comment by staticfloat** [comment]
> Should I add ∇logcosh(x) = tanh(x) gradient to Tracker?

For now, yes.  Eventually we'll need to add it to Zygote as well.

**Comment by thebhatman** [comment]
Is there anything else to be done here?

File: docs/src/training/training.md [hunk 4]
```diff
@@ -32,6 +32,18 @@ Flux.train!(loss, ps, data, opt)
 
 The objective will almost always be defined in terms of some *cost function* that measures the distance of the prediction `m(x)` from the target `y`. Flux has several of these built in, like `mse` for mean squared error or `crossentropy` for cross entropy loss, but you can calculate it however you want.
 
+In-built loss functions:
```
Comment (MikeInnes) [inline docs/src/training/training.md]: This should be in the API reference, not tutorial-style docs. Maybe in the layers section, if there's nowhere better. Good to have though.

**Comment by dhairyagandhi96** [comment]
Bump

File: src/layers/stateless.jl [hunk 5]
```diff
@@ -49,3 +49,29 @@ function normalise(x::AbstractArray; dims=1)
   σ′ = std(x, dims = dims, mean = μ′, corrected=false)
   return (x .- μ′) ./ σ′
 end
+
+function normalise(x::AbstractArray, dims)
```
Comment (MikeInnes) [inline src/layers/stateless.jl]: Is this intended to be part of this patch?

File: src/layers/stateless.jl [hunk 6]
```diff
@@ -49,3 +49,29 @@ function normalise(x::AbstractArray; dims=1)
...
+end
+
+"""
+    Poisson Loss function
+Poisson loss function is a measure of how the predicted distribution diverges from the expected distribution.
+https://isaacchanghau.github.io/post/loss_functions/
+"""
+poisson(ŷ, y) = sum(ŷ .- y .* log.(ŷ)) *1 // size(y,2)
+
+hinge(ŷ, y) = sum(max.(0, 1 .-  ŷ .* y)) *1 // size(y,2)
```
Comment (MikeInnes) [inline src/layers/stateless.jl]: `hinge` needs a doc string.

File: src/layers/stateless.jl (hunk 5 above)
Comment (thebhatman) [reply src/layers/stateless.jl]: Oops. Yeah I have removed it now.

File: docs/src/models/layers.md [hunk 7]
```diff
@@ -65,3 +65,15 @@ AlphaDropout
 LayerNorm
 GroupNorm
 ```
+
+## In-built loss functions:
```
Comment (MikeInnes) [inline docs/src/models/layers.md]:
```suggestion
## Loss Functions

```

File: src/layers/stateless.jl [hunk 8]
```diff
@@ -49,3 +49,28 @@ function normalise(x::AbstractArray; dims=1)
   σ′ = std(x, dims = dims, mean = μ′, corrected=false)
   return (x .- μ′) ./ σ′
 end
+
+"""
+    Kullback Leibler Divergence(KL Divergence)
```
Comment (MikeInnes) [inline src/layers/stateless.jl]: Could use a signature. Right now this title will show up as code.

File: src/layers/stateless.jl [hunk 9]
```diff
@@ -49,3 +49,28 @@ function normalise(x::AbstractArray; dims=1)
...
+"""
+    Poisson Loss function
+Poisson loss function is a measure of how the predicted distribution diverges from the expected distribution.
+https://isaacchanghau.github.io/post/loss_functions/
+"""
+poisson(ŷ, y) = sum(ŷ .- y .* log.(ŷ)) *1 // size(y,2)
+
+"""
+    Hinge Loss function
+Measures the loss given the prediction ŷ and true labels y(containing 1 or -1). This is usually used for measuring whether two inputs are similar or dissimilar
```
Comment (MikeInnes) [inline src/layers/stateless.jl]: Full stop at the end. The last sentence also doesn't sound that helpful right now (isn't that just the definition of a loss?); maybe it just needs rewording or a reference.

File: docs/src/models/layers.md [hunk 10]
```diff
@@ -66,7 +66,7 @@ LayerNorm
 GroupNorm
 ```
 
-## In-built loss functions:
+## Loss functions:
```
Comment (MikeInnes) [inline docs/src/models/layers.md]: The other headings in this section are capitalised, and don't end with a colon.

**Comment by thebhatman** [comment]
I have made the suggested changes in documentation, are there any more changes to be done?

**Comment by thebhatman** [comment]
Bump. Is there anything left to do here?

File: docs/src/models/layers.md [hunk 11]
```diff
@@ -65,3 +65,15 @@ AlphaDropout
 LayerNorm
 GroupNorm
 ```
+
+## Loss Functions
```
Comment (MikeInnes) [inline docs/src/models/layers.md]: I think we should perhaps label these "cost functions", since we sometimes refer to the whole objective function (including the model) as the "loss".

Comment (MikeInnes) [reply docs/src/models/layers.md]:
```suggestion
## Cost Functions
```

**Comment by MikeInnes** [comment]
Once you've changed the line I mentioned, feel free to merge.

bors delegate

**Comment by MikeInnes** [comment]
bors d+

**Comment by MikeInnes** [comment]
Is bors borked again? @dhairyagandhi96 do we have any idea how to fix it?

**Comment by dhairyagandhi96** [comment]
Seems like it, the gitlab yaml was changed recently, but I've noticed this with bors only recently. I'll take a look at that.

**Comment by dhairyagandhi96** [comment]
You might want to try again, updated some keys in gitlab.

**Comment by dhairyagandhi96** [comment]
Also, can you add a test with `Float32` and make sure no spurious promotions to `Float64` occur

**Comment by thebhatman** [comment]
Done. I have added the tests to check for spurious promotions.
